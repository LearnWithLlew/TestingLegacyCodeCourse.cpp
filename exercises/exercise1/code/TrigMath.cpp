#include "TrigMath.h"

double TrigMath::SinRaw(int idx) {
    return SIN_TABLE[idx & SIN_MASK];
}

double TrigMath::CosRaw(int idx) {
    return SIN_TABLE[(idx + COS_OFFSET) & SIN_MASK];
}

double TrigMath::Mxatan(double arg) {
    auto argsq = arg * arg;
    auto value = (((p4 * argsq + p3) * argsq + p2) * argsq + p1) * argsq + p0;
    value /= ((((argsq + q4) * argsq + q3) * argsq + q2) * argsq + q1) * argsq + q0;
    return value * arg;
}

double TrigMath::Msatan(double arg) {
    if (arg < sq2m1)
        return Mxatan(arg);
    if (arg > sq2p1)
        return HALF_PI - Mxatan(1 / arg);
    return HALF_PI / 2 + Mxatan((arg - 1) / (arg + 1));
}

void TrigMath::CreateSinTable() {
    for (auto i = 0; i < SIN_SIZE; i++)
        SIN_TABLE[i] = sin(i * TWO_PI / SIN_SIZE);
}

TrigMath::TrigMath() {
    CreateSinTable();
}

double TrigMath::Sin(double angle) {
    return SinRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

int TrigMath::Floor(double a) {
    return (int) a;
}

double TrigMath::Cos(double angle) {
    return CosRaw(Floor(angle * SIN_CONVERSION_FACTOR));
}

double TrigMath::Tan(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return SinRaw(idx) / CosRaw(idx);
}

double TrigMath::Sec(double angle) {
    return 1 / Cos(angle);
}

double TrigMath::Cot(double angle) {
    auto idx = Floor(angle * SIN_CONVERSION_FACTOR);
    return CosRaw(idx) / SinRaw(idx);
}

double TrigMath::Asin(double value) {
    if (value > 1)
        return NAN;
    if (value < 0)
        return -Asin(-value);
    auto temp = sqrt(1 - value * value);
    if (value > 0.7)
        return HALF_PI - Msatan(temp / value);
    return Msatan(value / temp);
}

double TrigMath::Acos(double value) {
    if (value > 1 || value < -1)
        return NAN;
    return HALF_PI - Asin(value);
}

double TrigMath::Atan(double value) {
    if (value > 0)
        return Msatan(value);
    return -Msatan(-value);
}

double TrigMath::Atan2(double y, double x) {
    if (y + x == y)
        return y >= 0 ? HALF_PI : -HALF_PI;
    y = Atan(y / x);
    if (x < 0) {
        if (y <= 0) {
            return y + PI;
        } else {
            return y - PI;
        }
    }
    return y;
}

double TrigMath::Acsc(double value) {
    if (value == 0)
        return NAN;
    return Asin(1 / value);
}

double TrigMath::Asec(double value) {
    if (value == 0)
        return NAN;
    return Acos(1 / value);
}

double TrigMath::Acot(double value) {
    if (value == 0)
        return NAN;
    if (value > 0)
        return Atan(1 / value);
    return Atan(1 / value) + PI;
}

