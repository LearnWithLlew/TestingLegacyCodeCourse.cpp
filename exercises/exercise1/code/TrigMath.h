//
// Created by LLEWELLYN FALCO on 12/10/17.
//

#ifndef CATCHPLAYGROUND_TRIGMATH_H
#define CATCHPLAYGROUND_TRIGMATH_H

#define _USE_MATH_DEFINES
#include <math.h>
#include <cmath>

class TrigMath {
public:
    const double PI = M_PI;
    const double SQUARED_PI = PI * PI;
    const double HALF_PI = PI / 2;
    const double QUARTER_PI = HALF_PI / 2;
    const double TWO_PI = 2 * PI;
    const double THREE_PI_HALVES = TWO_PI - HALF_PI;
    const double DEG_TO_RAD = PI / 180;
    const double HALF_DEG_TO_RAD = PI / 360;
    const double RAD_TO_DEG = 180 / PI;
    double SQRT_OF_TWO = std::sqrt(2);
    double HALF_SQRT_OF_TWO = SQRT_OF_TWO / 2;
private  :

    const int SIN_SIZE = 100000;
    double SIN_TABLE[100000];
    const int SIN_MASK = SIN_SIZE - 1;
    const double SIN_CONVERSION_FACTOR = SIN_SIZE / TWO_PI;

    const int COS_OFFSET = SIN_SIZE / 4;

//Arc trig
    const double sq2p1 = 2.414213562373095048802;

    const double sq2m1 = 0.414213562373095048802;
    const double p4 = 0.161536412982230228262E2;
    const double p3 = 0.26842548195503973794141E3;
    const double p2 = 0.11530293515404850115428136E4;
    const double p1 = 0.178040631643319697105464587E4;
    const double p0 = 0.89678597403663861959987488E3;
    const double q4 = 0.5895697050844462222791E2;
    const double q3 = 0.536265374031215315104235E3;
    const double q2 = 0.16667838148816337184521798E4;
    const double q1 = 0.207933497444540981287275926E4;
    const double q0 = 0.89678597403663861962481162E3;

    double sinRaw(int idx);

    double cosRaw(int idx);

    double mxatan(double arg);

    double msatan(double arg);

    void createSinTable();

public:

    TrigMath();


    double sin(double angle);

    int floor(double a);


    double cos(double angle);

    double tan(double angle);

    double csc(double angle) {
        return 1 / sin(angle);
    }

    double sec(double angle);


    double cot(double angle);

    double asin(double value);

    double acos(double value);

    double atan(double value);

    double atan2(double y, double x);

    double acsc(double value);

    double asec(double value);

    double acot(double value);


};

#endif //CATCHPLAYGROUND_TRIGMATH_H
