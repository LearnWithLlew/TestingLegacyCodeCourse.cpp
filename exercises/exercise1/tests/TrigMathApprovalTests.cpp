#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"
#include "TrigMath.h"

using namespace ApprovalTests;
using namespace std;


//TEST_CASE("ApprovalTest Sin") {
//    TrigMath math;
//	Approvals::verify(math.Sin(0.5));
//}
//
//TEST_CASE("ApprovalTest Cos") {
//	TrigMath math;
//	Approvals::verify(math.Cos(0.5));
//}

std::string runEverything(double input) {
	TrigMath math;
	std::stringstream outputStream;
	outputStream << "Sin: " << math.Sin(input)<< "\n";
	outputStream << "Cos: " << math.Cos(input) << "\n";
	outputStream << "Acos: " << math.Acos(input) << "\n";
	outputStream << "Asin: "<< math.Asin(input) << "\n";
	outputStream << "Acot: " << math.Acot(input) << "\n";
	outputStream << "Acsc: " << math.Acsc(input) << "\n";
	outputStream << "Asec: " << math.Asec(input) << "\n";
	outputStream << "Atan: " << math.Atan(input) << "\n";
	outputStream << "Atan2: " << math.Atan2(input, input) << "\n";
	outputStream << "Cot: " << math.Cot(input) << "\n";
	outputStream << "Csc: " << math.Csc(input) << "\n";
	outputStream << "Sec: " << math.Sec(input) << "\n";
	outputStream << "Tan: " << math.Tan(input) << "\n";

	return outputStream.str();
}

//TEST_CASE("ApprovalTest run everything") {
//	Approvals::verify(runEverything(1.0));
//}
//
//TEST_CASE("ApprovalTest run everything with pi") {
//	Approvals::verify(runEverything(3.14));
//}

TEST_CASE("ApprovalTest run everything combined")
{
	std::vector<double> inputs{ 3.14, 1.0, -0.1, 0, 0.2, 0.8 };
	Approvals::verifyAll("Combination test", inputs, [](auto input, auto& stream) {
		stream << input << " =>\n" << runEverything(input);
			
		});
}

TEST_CASE("CombinationsStartingPoint")
{
	TrigMath math;
	std::vector<double> inputs1{ 3.14, 1.0, -0.1, 0, 0.2, 0.8 };
	std::vector<double> inputs2{ 3.14, 1.0, -0.1, 0, 0.2, 0.8 };
	
	CombinationApprovals::verifyAllCombinations(
		[&](auto input1, auto input2) { return math.Atan2(input1, input2); }, inputs1, inputs2);
}
