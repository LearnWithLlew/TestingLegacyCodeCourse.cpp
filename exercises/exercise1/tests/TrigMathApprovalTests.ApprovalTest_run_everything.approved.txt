Sin: 0.0970479
Cos: 0.828807
Acos: 0
Asin: 1.5708
Acot: 0.785398
Acsc: 1.5708
Asec: 0
Atan: 0.785398
Atan2: 0.785398
Cot: 8.54019
Csc: 10.3042
Sec: 1.20655
Tan: 0.117093

